src/wallet3/transaction_constructor.cpp

```diff
@@ -1,13 +1,103 @@
+    PendingTransaction txNew(recipients);
+    SelectInputsAndFinalise(txNew);
+    return txNew;
+  }
+  
+  // SelectInputs will choose some available unspent outputs from the database and allocate to the transaction
+  // can be called multiple times and will add until enough is sufficient
+  void
+  TransactionConstructor::SelectInputs(PendingTransaction& ptx) const
+  {
+    // Fail early
+    int64_t estimated_fee = EstimateFee();
+    //int64_t estimated_fee = estimate_fee(2, fake_outs_count, min_outputs, extra.size(), clsag, base_fee, fee_percent, fixed_fee,     fee_quantization_mask);
+    int64_t transaction_total = ptx.SumOutputs() + estimated_fee;
+    int64_t wallet_balance = db->prepared_get<int>("SELECT sum(amount) FROM outputs");
+
+    // Check that we actually have enough in the outputs to build this transaction.
+    if (wallet_balance < transaction_total)
+      throw std::runtime_error("Insufficient Wallet Balance");
+
+
+    int64_t shortfall = transaction_total - ptx.SumInputs();
+
+    // If we already have enough inputs return
+    if (shortfall < 0)
+      return;
+
+    // Prefer a single output if suitable
+    SQLite::Statement single_output{db->db, "SELECT * FROM outputs WHERE amount > ? ORDER BY amount ASC LIMIT 1"};
+    single_output.bind(1, shortfall);
+    while (single_output.executeStep())
+    {
+      wallet::Output o{};
+      o.amount       = single_output.getColumn(1).getInt64();
+      o.output_index = single_output.getColumn(2).getInt64();
+      o.global_index = single_output.getColumn(3).getInt64();
+      o.unlock_time  = single_output.getColumn(4).getInt64();
+      o.block_height = single_output.getColumn(5).getInt64();
+      o.block_time   = single_output.getColumn(6).getInt64();
+      o.spending     = static_cast<bool>(single_output.getColumn(7).getInt64());
+      o.spent_height = single_output.getColumn(8).getInt64();
+      o.spent_time   = single_output.getColumn(9).getInt64();
+      shortfall -= o.amount;
+      ptx.chosenOutputs.push_back(o);
+      ptx.UpdateChange();
+      return;
+    }
+
+    // Else select some random outputs
+    // TODO sean amount > dust to prevent increasing fee higher than input amount
+    SQLite::Statement many_outputs{db->db, "SELECT * FROM outputs ORDER BY amount"};
+    while (shortfall > 0)
+    {
+      if (many_outputs.executeStep()) {
+        wallet::Output o{};
+        o.amount       = many_outputs.getColumn(1).getInt64();
+        o.output_index = many_outputs.getColumn(2).getInt64();
+        o.global_index = many_outputs.getColumn(3).getInt64();
+        o.unlock_time  = many_outputs.getColumn(4).getInt64();
+        o.block_height = many_outputs.getColumn(5).getInt64();
+        o.block_time   = many_outputs.getColumn(6).getInt64();
+        o.spending     = static_cast<bool>(many_outputs.getColumn(7).getInt64());
+        o.spent_height = many_outputs.getColumn(8).getInt64();
+        o.spent_time   = many_outputs.getColumn(9).getInt64();
```

Comment:
I would suggest making a wallet::Output constructor that takes a `std::tuple<int64_t, int64_t, ...., bool, int64_t, int64_t>` argument, then you can do something like:
```c++
wallet::Output o{db::get<int64_t, int64_t, ....., bool, int64_t>(single_output)};
```
and have that constructor set all the fields, rather than the extraction duplication here.

Reply:
Yeah nice!